raft/src/main/java/org/apache/kafka/snapshot/SnapshotWriter.java

```diff
@@ -78,6 +83,76 @@ public SnapshotWriter(
         );
     }
 
+    /**
+     * Adds a {@link MetadataSnapshotHeaderRecord} to snapshot
+     *
+     * @throws IllegalStateException if the snapshot is not empty
+     */
+    private void initializeSnapshotWithHeader() {
+        if (snapshot.sizeInBytes() != 0) {
+            throw new IllegalStateException("Initializing new snapshot (ID: "
+                    + snapshot.snapshotId().epoch + ", " + snapshot.snapshotId().offset
+                    + ") with a non-empty file");
+        }
+        MetadataSnapshotHeaderRecord headerRecord = new MetadataSnapshotHeaderRecord()
+            .setVersion(ControlRecordUtils.METADATA_SNAPSHOT_HEADER_VERSION);
+        accumulator.appendSnapshotHeaderMessage(headerRecord);
+        accumulator.forceDrain();
+    }
+
+    /**
+     * Add a {@link MetadataSnapshotFooterRecord} to the snapshot
+     *
+     * No more records should be appended to the snapshot after calling this method
+     *
+     * @throws IllegalStateException if the snapshot is empty (no header)
+     */
+    private void finalizeSnapshotWithFooter() {
+        if (snapshot.sizeInBytes() == 0) {
```

Comment:
Hmm. Not sure but I would argue that this check is always false since there is only once "constructor" and that always appends at least one record batch. We can confirm in the tests that this is always true.

Reply:
I agree with the argument, and adding a test, but do you think we should remove the code check? For me the code check serves as a runtime fail safe and also a hint to the reader. The check is in no way complete. I do not have a strong opinion here. Just thinking out loud.